src/ql/arch/cc/pass/gen/vq1asm/detail/codegen.h

```diff
@@ -25,123 +29,185 @@ namespace gen {
+     * Our strategy is to first process all CustomInstruction's in a bundle, storing the relevant information in
+     * BundleInfo. Then, when all work for a bundle has been collected, we generate code in bundle_finish
+     */
+
+    /**
+     * Clear bundleInfo, which maintains the work that needs to be performed for bundle
+     */
+    void bundle_start(const Str &cmnt);
+
+    /**
+     * Generate code for bundle from information collected in bundleInfo (which may be empty if no custom gates are
+     * present in bundle)
+     */
+    void bundle_finish(UInt startCycle, UInt durationInCycles, Bool isLastBundle);
+
+    /**
+     * Collect information from CustomInstruction (single/two/N qubit gate, including readout. FKA as gate). Translates
+     * 'gate' representation to 'waveform' representation (BundleInfo) and maps qubits to instruments & group.
+     *
+     * Does not deal with the control mode and digital interface of the instrument, since we first need to collect all
+     * work per instrument
+     */
+    void custom_instruction(const ir::CustomInstruction &custom);
+
+    // Structured control flow
+    void if_elif(const ir::ExpressionRef &condition, const Str &label, Int branch);
+    void if_otherwise(const Str &label, Int branch);
+    void if_end(const Str &label);
+    void foreach_start(const ir::Reference &lhs, const ir::IntLiteral &frm, const Str &label);
+    void foreach_end(const ir::Reference &lhs, const ir::IntLiteral &frm, const ir::IntLiteral &to, const Str &label);
+    void repeat(const Str &label);
+    void until(const ir::ExpressionRef &condition, const Str &label);
+    void for_start(utils::Maybe<ir::SetInstruction> &initialize, const ir::ExpressionRef &condition, const Str &label);
+    void for_end(utils::Maybe<ir::SetInstruction> &update, const Str &label);
+    void do_break(const Str &label);
+    void do_continue(const Str &label);
 
     void comment(const Str &c);
 
+    /**
+     * Handle expressions.
+     *
+     * To understand how cQASM functions end up in the IR, please note that functions are handled during analyzing
+     * cQASM, see 'AnalyzerHelper::analyze_function()'.
+     *
+     * A default set of functions that only handle constant arguments is provided by libqasm, see
+     * 'register_into(resolver::FunctionTable &table)'. These functions add a constant node to the IR when called (and
+     * fail if the arguments are not constant)
+     *
+     * Some of these are overridden by OpenQL to allow use of non-constant arguments. This is a 2 step process, where
+     * 'convert_old_to_new(const compat::PlatformRef &old)' adds functions to ir->platform using 'add_function_type',
+     * and 'ql::ir::cqasm:read()' then walks 'ir->platform->functions' and adds the functions using
+     * 'register_function()'. These functions add a 'cqv::Function' node to the IR (even if the arguments are constant,
+     * so overriding a function defeats libqasm's constant removal for that function).
```

Comment:
Is this a bug?

Reply:
No (otherwise I would have added a "FIXME" ;-), it was designed this way, the comment is just there because it's not immediately apparent that things work this way (it took some emailing with Jeroen for me to grasp it). 

Note that https://github.com/QuTech-Delft/OpenQL/issues/424 contains ideas to make function handling more flexible (among many other things)